core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1844,6 +1900,68 @@ private Optional<OutputDataSizeEstimate> getEstimatedOutputDataSize()
             return Optional.of(new OutputDataSizeEstimate(estimateBuilder.build()));
         }
 
+        private Optional<OutputDataSizeEstimate> getEstimatedSmallStageOutputDataSize(Function<StageId, StageExecution> stageExecutionLookup)
+        {
+            if (!smallStageEstimationEnabled) {
+                return Optional.empty();
+            }
+
+            if (smallStageRequireNoMorePartitions && !isNoMorePartitions()) {
+                return Optional.empty();
+            }
```

Comment:
We should check that the small stage has made some progress, right?

Reply:
hmm - yeah - I guess it is a good point. At laest one of sources - right?

Reply:
Yeah, the progress has to be non-zero